File: tournament_winner/tournament_winner.py
# Complexity O(n) n - number of competitions,
# Space O(k) k - number of teams
def tournamentWinner(competitions, results):
    winners = {}
    winner = ""
    for i, teams in enumerate(competitions):
        team = teams[0] if results[i] else teams[1]
        if team in winners:
            winners[team] += 3
        else:
            winners[team] = 3
        if winner != "" and winners[team] > winners[winner]:
            winner = team
        if winner == "":
            winner = team
    return winner


competitions = [['HTML', 'C#'], ['C#', 'Python'], ['Python', 'HTML']]
results = [0, 0, 1]

if __name__ == '__main__':
    print(tournamentWinner(competitions, results))

File: tournament_winner/tournamentWinner.go
package main

import "fmt"

func tournamentWinner(competitions *[][2]string, results *[]int) string {
	var winner string
	winners := make(map[string]int)

	for i, teams := range *competitions {
		team := teams[1-(*results)[i]]
		if _, found := winners[team]; found {
			winners[team] += 3
		} else {
			winners[team] = 3
		}
		if winner != "" {
			if winners[team] > winners[winner] {
				winner = team
			}
		} else {
			winner = team
		}
	}
	return winner
}

var competitions = [][2]string{{"HTML", "C#"}, {"C#", "Python"}, {"Python", "HTML"}}
var results = []int{0, 0, 1}

func main() {
	fmt.Println(tournamentWinner(&competitions, &results))
}
